### PR TITLE
Dynamic grid blocks: Try using ServerSideRender component for preview

### PR DESCRIPTION
**Pull request by ryelle**
See https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/558#discussion_r283286010 – In that PR, the blocks are dynamic - rendered in PHP but previewed with a JS component. It would be nice if we could keep all the rendering in one place. With the `ServerSideRender` component, we can call an API endpoint which returns the `render_callback` content given the attributes passed in.

This also means we can remove the code for fetching products from the API endpoint, and let the PHP callback do all the work.

I've marked this as in progress, but would like some feedback on the approach before I keep going.

#### Screenshots

Currently this makes for an ugly preview, but it _is_ pulling from the PHP function. We might need to add CSS to the editor, or tweak the HTML in the PHP function… either way this needs work.

Screen Shot 2019-05-14 at 5 36 50 PM

#### Accessibility

Note, this PR also wraps the content in `Disabled`, which prevents the links/buttons from being clickable. Before merging, we'll want to run through and see how this behaves with a screen reader.

### How to test the changes in this Pull Request:

1. Add a Newest Products or On Sale Products block
2. It should preview correctly, with the expected products/order/etc
3. Publish your post
4. View it, it should have the same (correct) products in the same order.

### CODE REVIEW

**Comment by mikejolley** [comment]
First impressions, it's nice that this shares the same PHP based templates. That would also ensure things like filters are respected.

That said, I see the drawback here that rendering is much much slower and see now why the documentation has _Server-side render is meant as a fallback; client-side rendering in JavaScript is always preferred (client rendering is faster and allows better editor manipulation)._

When testing the editing view didn't update when toggling things on/off. It did work on frontend and saved correctly. You're probably already aware but just making a note.

**Comment by ryelle** [comment]
Update: fixed the content visibility issue, and tweaked the CSS a bit for better previewing. I also played with loading in storefront's `woocommerce.css` to get some more of the front-end CSS included in the editor, and it looks pretty good:

565-with-woo-css

And this is the new default (what you would get if you run this PR)

565-default

It would be nice if the loading placeholder showed which block was loading, otherwise when you load the editor with multiple blocks you get [a bunch of spinners, and no other info.](https://user-images.githubusercontent.com/541093/57808633-c55fbf00-7731-11e9-90c5-cc90bb6e35e2.png) There's another ServerSideRender PR in progress that'll move the component to a new package, once that's done I'll open a PR to add in the block title & icon to see if that's something we want.

> That said, I see the drawback here that rendering is much much slower

yeah, there's definitely trade-offs here…

**Benefits**

- Uses the same template for front-end and editor, easier for us to maintain
- Same markup in editor & front-end will be easier for themes to style
- Using PHP to render the callback means we could introduce new block hooks for other devs

**Drawbacks**

- No longer a smooth, "live update" feel (looses some of that gutenberg magic ✨)
- Technically doing more API calls (every setting change calls the API, before we skipped calling the API on visual changes)
- It might be confusing later that some blocks are `ServerSideRender`ed and some are using JS components (like Featured Product, new blocks that haven't been written yet?)

I _think_ the benefits outweigh the drawbacks in the case of the grid blocks, but I'm not sure if this is really a pattern we want to establish for the rest of the blocks – whatever we pick here we'll probably use for the Reviews blocks too – or not, the "Load More" functionality there is going to throw a wrench into things.

We could always swing in the other direction and use the JS components on the front-end, I believe that's what Bookings ended up doing… I don't think I'd actually recommend that, but it's a possibility.

File: assets/css/editor.scss [hunk 1]
```diff
@@ -28,6 +35,43 @@
 		display: block;
 	}
 
+	.star-rating {
```
Comment (mikejolley) [inline assets/css/editor.scss]: This is fine for now, but I raised the point here https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/576#issuecomment-494338138

Comment (ryelle) [reply assets/css/editor.scss]: Yeah, if we go with your "option 1", we'll end up moving this to style.css, but we can handle that then 👍

**Pull request merged**